python/ray/serve/scripts.py

```diff
@@ -284,70 +287,54 @@ def run(
     blocking: bool,
     gradio: bool,
 ):
-    sys.path.insert(0, app_dir)
+    # If no address is given and no local ray instance is running, we want to start one.
+    if address is None:
+        ray.init(namespace=SERVE_NAMESPACE)
 
     final_runtime_env = parse_runtime_env_args(
         runtime_env=runtime_env,
         runtime_env_json=runtime_env_json,
         working_dir=working_dir,
     )
+    if "env_vars" not in final_runtime_env:
+        final_runtime_env["env_vars"] = {}
+    final_runtime_env["env_vars"]["RAY_JOB_STOP_SIGNAL"] = "SIGINT"
+
+    # The job to run on the cluster, which imports and runs the serve app.
+    from ray.serve import run_script
+
+    with open(run_script.__file__, "r") as f:
+        script = f.read()
+
+    # Use Ray Job Submission to run serve.
+    client = JobSubmissionClient(address)
+    submission_id = client.submit_job(
+        entrypoint=(
+            f"python -c '{script}' "
+            f"--config-or-import-path={config_or_import_path} "
+            f"--host={host} "
+            f"--port={port} "
+            f"--app-dir={app_dir} "
+            + ("--blocking " if blocking else "")
+            + ("--gradio " if gradio else "")
+        ),
+        # Setting the runtime_env will set defaults for the deployments.
+        runtime_env=final_runtime_env,
+    )
 
-    if pathlib.Path(config_or_import_path).is_file():
-        config_path = config_or_import_path
-        cli_logger.print(f'Deploying from config file: "{config_path}".')
+    async def print_logs():
+        async for lines in client.tail_job_logs(submission_id):
+            print(lines, end="")
 
-        with open(config_path, "r") as config_file:
-            config = ServeApplicationSchema.parse_obj(yaml.safe_load(config_file))
-        is_config = True
-    else:
-        import_path = config_or_import_path
-        cli_logger.print(f'Deploying from import path: "{import_path}".')
-        node = import_attr(import_path)
-        is_config = False
-
-    # Setting the runtime_env here will set defaults for the deployments.
-    ray.init(address=address, namespace=SERVE_NAMESPACE, runtime_env=final_runtime_env)
-
-    if is_config:
-        client = _private_api.serve_start(
-            detached=True,
-            http_options={
-                "host": config.host,
-                "port": config.port,
-                "location": "EveryNode",
-            },
-        )
-    else:
-        client = _private_api.serve_start(
-            detached=True,
-            http_options={"host": host, "port": port, "location": "EveryNode"},
-        )
+    def interrupt_handler():
+        # Upon keyboard interrupt, stop job (which sends an interrupt signal to the job
+        # and shuts down serve). Then continue to stream logs until the job finishes.
```

Comment:
Could we add a debug log statement here saying something like:

```
logger.debug("Received an interrupt locally. Forwarding interrupt to Serve app.")
```

If Serve apps aren't shutting down correctly, this could help us tell whether the issue is happening locally or on the remote cluster.

Reply:
Resolved offline. Adding a log statement here does not make it show up in the Ray logs since a Ray cluster isn't started locally.

Reply:
+1 on this, do we show any logs in local indicating receiving the interrupt signal and shutting down the cluster? @zcin

Reply:
There are logs on the cluster side when the submitted job receives an interrupt signal from `stop_job`, and those logs are streamed to the local console.